### ISSUE TITLE
Code sections cannot capture arguments when not using K&R style braces

### ISSUE THREAD

**Issue by ShadauxCat** [daokoder/dao#379]
This works:

```
s = string(5) { [index]
    (string) index;
}
```

This does not:

```
s = string(5)
{ [index]
    (string) index;
}
```

This also does not:

```
s = string(5)
{
    [index]
    (string) index;
}
```

Second and third both result in the same error:

```
[[ERROR]] in file "code string":
  At line 2 : Invalid expression --- " [index] ";
  At line 2 : Invalid expression --- " index] ";
  At line 2 : Symbol not defined --- " index ";
```

**Comment by ShadauxCat** [comment]
Actually on second thought, I suspect this is actually an issue that code sections are not recognized at all if the curly brace is on a different line. Supported by the fact that this:

```
s = string(5) {
    [index]
    (string) index;
}
```

Works.

I'ma change the title.

**Comment by ShadauxCat** [comment]
If parsing without requiring K&R style braces is overly difficult due to the confusion of whether
```
s = string(5)
{
    [index]
    (string) index;
}
```
Means 
```
s = string(5);

{
    [index]
    (string) index;
}
```

Then I would suggest either: 1) Make semicolons non-optional, or 2) allow the following syntax:

```
s = string(5) \
{
    [index]
    (string) index;
}
```
(note the backslash)

I know I'm not the only programmer in the world who strongly dislikes K&R style braces. I've worked with a lot of others who agree. Forcing K&R style braces isn't ideal, to say the least.

**Comment by dumblob** [comment]
It's a bug:
~~~~ruby
s = string(5)
{
    [index]
    (string) index;
}
~~~~
should be parsed the same way as
~~~~ruby
s = string(5) { [index]
    (string) index;
}
~~~~
The programmer has always an opportunity to put `;` there to force the parser to split it. No backslashes nor required semicolons are needed.

**Comment by Night-walker** [comment]
It's not really about style, and **it's not a bug**. As you've noticed, `string(5)` and `{...}` are treated as separate statements. This happens because the parser continues the current statement after a newline only if the current line ends with a binary operator, or the next line starts with a dot.

Such rules allow to make semicolons optional, and I would prefer them to stay that way -- semicolons are mostly nothing but a visual noise. Since braces may mean a list in Dao, there is no simple way to disambiguate it with a code section.

Backslashing may be added, but I'm not sure it's worth the fuss. Perhaps using a macro would be a better option.

**Comment by dumblob** [comment]
>Since braces may mean a list in Dao, there is no simple way to disambiguate it with a code section.

In the demonstrated case, it's not ambiguous - it can't be a list. That's why I called it a bug.

**Comment by Night-walker** [comment]
> In the demonstrated case, it's not ambiguous - it can't be a list.

But the parser doesn't know that in general case. If you omit things like `[...]` and `;` within the braces, it is no different from a list. But you may be right in that it might be possible to distinguish sections when they start with `[...]` -- if the parser can deduce that it's not an array.

**Comment by dumblob** [comment]
Even without deep investigation (`[...]` etc.) - it can't be anything else than a code section in this case (assuming *dead code* lists are not supported). If I wanted it to be a list, I'd write
~~~~ruby
s = string(5) ;  # semicolon
{
    [index]
    (string) index;
}
.iterate { io.writeln("I swear, I'll never ever write it unreadable again.", X) }
~~~~

**Comment by Night-walker** [comment]
It might not be trivial to check in advance what you imply. Furthermore, a list which is not used is not necessarily a dead code -- it's items may be constructed using non-pure functions.

**Comment by dumblob** [comment]
For such cases, we have the semicolon :wink:

**Comment by ShadauxCat** [comment]
I tend to agree with dumblob, but even in the case that Night-walker's putting forward, I would still request the ability to achieve this with a backslash before the newline. Being forced into K&R style for code sections is sub-ideal.

**Comment by dumblob** [comment]
>I would still request the ability to achieve this with a backslash before the newline.

Why? My proposal suggests that it's always a code section and therefore you can write it in non-K&R style without backslash.

**Comment by ShadauxCat** [comment]
No no no. I agree with your proposal, @dumblob. @Night-walker seems to think it won't work, though. If your proposal WILL work (and honestly I think it would) then it would be my first choice.

I'm just saying if there's a reason that it *won't* work, I would at least want the backslash option to be present. It's better than nothing.

**Comment by Night-walker** [comment]
It will work if the parser is told to look ahead for possible `{` after a newline and treat it *a priori* as a code section. But that also implies the need for extra `;` in certain cases.

**Comment by ShadauxCat** [comment]
Maybe it's because I'm primarily a C++ programmer, but the need for a semicolon to disambiguate the two cases doesn't bother me much.
